src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c

```diff
@@ -5180,31 +5184,236 @@ int wm_vuldet_collect_agent_software(scan_agent *agent, sqlite3 *db, scan_ctx_t*
+    while(agents_array && agents_array[i] != OS_INVALID && current_task && WM_TASK_CANCELLED != current_task->task_status) {
+        if (wm_vuldet_collect_agent_info(agents_array[i], vuldet)) {
+            mterror(WM_VULNDETECTOR_LOGTAG, VU_NO_AGENT_ERROR);
+        } else {
+            snprintf(status_message, OS_SIZE_64, " Agent: %.3d.", agents_array[i]);
+            wm_vuldet_create_and_send_message(current_task, WM_TASK_IN_PROGRESS, status_message);
+            if (wm_vuldet_run_scan(vuldet)) {
+                mterror(WM_VULNDETECTOR_LOGTAG, VU_AG_CHECK_ERR);
+            }
+        }
+        i++;
+        current_task = (vu_task_context_t*) linked_queue_read_ex(vu_tasks_queue);
+    }
+    mtinfo(WM_VULNDETECTOR_LOGTAG, VU_END_SCAN);
 
-        if ((curr_time - vuldet->last_scan >= vuldet->scan_interval)) {
-            wm_vuldet_run_scan(vuldet);
+    wm_vuldet_create_and_send_message(current_task, WM_TASK_DONE, NULL);
+    os_free(agents_array);
+}
+
+void wm_vuldet_working_thread(void* vuldet_data) {
+    wm_vuldet_t* vuldet = (wm_vuldet_t*) vuldet_data;
+    vu_task_context_t* current_task = NULL;
+
+    while (wm_vuldet_working_thread_active) {
+        current_task = (vu_task_context_t*) linked_queue_read_ex(vu_tasks_queue);
+        if (current_task) {
+            switch (current_task->task_type) {
+                case VU_ON_DEMAND_SCAN:
+                    wm_vuldet_on_demand_scan(current_task, vuldet);
+                    break;
+                case VU_ON_DEMAND_FEED_UPDATE:
+                    wm_vuldet_on_demand_feed_update(current_task, vuldet);
+                    break;
+                case VU_INTERVAL_FEED_UPDATE:
+                    wm_vuldet_feed_update_by_interval(current_task, vuldet);
+                    break;
+                case VU_INTERVAL_SCAN:
+                    wm_vuldet_scan_by_interval(current_task, vuldet);
+                    break;
+                default:
+                    mdebug1("Invalid task type for vulnerability detector.");
+                    break;
+            }
+            // Removing task from queue
+            vu_tasks_task_context_free((vu_task_context_t*) linked_queue_pop_ex(vu_tasks_queue));
+        } else {
+            sleep(1);
         }
+    }
+}
+
+int wm_vuldet_run_working_thread(pthread_t* working_thread, wm_vuldet_t* vuldet) {
+    int status;
+
+    if (status = pthread_create(working_thread, NULL, (void*)&wm_vuldet_working_thread, (void*) vuldet), status != 0) {
+        mwarn("Couldn't create vulnerability detector working thread: %s", strerror(status));
+        return OS_INVALID;
+    }
+
+    return OS_SUCCESS;
+}
+
+void *wm_vuldet_main(wm_vuldet_t* vuldet) {
+    wm_vuldet_init(vuldet);
+
+    wm_vuldet_check_db();
 
-        wm_vuldet_run_sleep(vuldet);
+    if(OS_SUCCESS == wm_vuldet_init_queue()) {
+        pthread_t working_thread;
+        if (OS_SUCCESS == wm_vuldet_run_working_thread(&working_thread, vuldet)) {
+            while (TRUE) {
```

Comment:
I would stop using this "TRUE" definition and start using boolean.

Reply:
Done